File: cs50x/Problem Set 6/sentimental-credit/credit.py
from cs50 import get_int
from math import ceil

def main():
    cardNum = get_int('Input card number: ')
    cardNum = str(cardNum)
    lenght = len(cardNum)
    if algorithm(cardNum):
        if lenght == 13 or lenght == 16 and cardNum[0] == '4':
            print('VISA')
        elif lenght == 15 and cardNum[0] == '3' and (cardNum[1] == '4' or cardNum[1] == '7'):
            print('AMEX')
        elif lenght == 16 and cardNum[0] == '5' and (cardNum[1] in '12345'):
            print('MASTERCARD')
        else:
            print('INVALID')
    else:
        print('INVALID')

def algorithm(cardNum):
    i = 2
    j = 1
    n = int(len(cardNum) / 2)
    secondN = ceil(len(cardNum) / 2)
    sum = 0

    for x in range(n):
        rest = 0
        penultimate = cardNum[len(cardNum) -i]
        luhn = int(penultimate) * 2
        if luhn > 9:
            rest = luhn % 10
            luhn = 1
        sum += luhn + rest
        i += 2

    for y in range(secondN):
        last = cardNum[len(cardNum) -j]
        sum += int(last)
        j += 2

    if sum % 10 == 0:
        return True
    else:
        return False

if __name__ == '__main__':
    main()

File: cs50x/Problem Set 1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int sum1(long n);
int sum2(long n);
int number_lenght(long n);
int first_number(long n);

int main(void)
{
    int checksum;
    int lenght;
    int first;
    long n;

    do
    {
    n = get_long("Enter credit card number: ");
    }
    while(n < 0);

    checksum = sum1(n) + sum2(n);
    lenght = number_lenght(n);
    first = first_number(n);

    if(checksum % 10 == 0 && lenght == 15 && (first == 34 || first == 37))
    {
    printf ("AMEX\n");
    }

    else if(checksum % 10 == 0 && lenght == 16 && (first > 50 && first < 56))
    {
    printf ("MASTERCARD\n");
    }

    else if(checksum % 10 == 0 && (lenght == 13 || lenght == 16) && (first > 39 && first < 50))
    {
    printf ("VISA\n");
    }

    else
    {
        printf ("INVALID\n");
    }
}

// checksum function first part
int sum1(long n)
{
    int sum=0;

     while(n > 1)
    {
    sum = n % 10 + sum;
    n = n / 100;
    }
    return sum;
}


// checksum function second part
int sum2(long n)
{
    int broj;
    int sum;
    int dvoznamenka;
    n = n / 10;

    while(n > 1)
    {
    broj = n % 10;
    if(broj < 5)
    {
        sum = broj * 2 + sum;
    }
    else
    {
        dvoznamenka = broj * 2;
        sum = dvoznamenka / 10 + sum;
        sum = dvoznamenka % 10 + sum;
    }
    n = n / 100;
    }

    return sum;
}

int number_lenght(long n)
{
    int lenght = 0;

    while(n > 1)
    {
    n = n / 10;
    lenght++;
    }
    return lenght;
}

int first_number(long n)
{
    while(n >= 100)
    {
        n = n / 10;
    }
    return n;
}
